### PR TITLE
[api-proposal] api proposal first draft

### PR DESCRIPTION
**Pull request by sorahn**

### CODE REVIEW

File: api-proposal.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+- GET `/api/v1/package/:package`
+  ```
+  data: <package>
+  ```
+- POST `/api/v1/package`
+  ```
+  data: {
+    package: <package>
+    releases: [<release>]
```
Comment (sorahn) [inline api-proposal.md]: This would probably be an array of `id`'s.

File: api-proposal.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    tags: [<tag>]
```
Comment (sorahn) [inline api-proposal.md]: If tags are just strings, and duplicates are not allowed, then this can just be the array of tags, no need for IDs or anything.

**Comment by sorahn** [comment]
> Let's discuss if this endpoint is needed.
> Instead of `/api/v1/categories/connector`, you would just do `/api/v1/search?query=tag:connector`?
> This way, the backend doesn't need to be aware of the concept of categories at all

So are the categories just special tags, or would we actually differentiate them? I *think* it might be easier if we just have `cat:connector` vs `tag:something`, and then we could probably maintain a cat<->package relation somewhere.

File: api-proposal.md (hunk 1 above)
Comment (rmetzger) [reply api-proposal.md]: I agree.

File: api-proposal.md (hunk 2 above)
Comment (rmetzger) [reply api-proposal.md]: +1

File: api-proposal.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+- GET `/api/v1/tags`
+  ```
+  items: [<tag>]
+  ```
+
+---
+
+### Authentication
+
+Is the user authenticated?
```
Comment (klion26) [inline api-proposal.md]: I think here we can add more information if needed?

File: api-proposal.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  }
+  ```
+- POST `/api/v1/package`
+  ```
+  data: <package>
+  ```
+- PATCH `/api/v1/package/:package`
+  ```
+  data: <package>
+  ```
```
[the remark this answers is not in the document]
Comment (sorahn) [reply api-proposal.md]: Ah, yeah you're right.  I missed that.  Good catch.

File: api-proposal.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+#### Package Releases
+
+- **Schema**
+
+  ```
+  <release> {
+    id: number
+    number: string //(semver probably)
+    description: string
```
[the remark this answers is not in the document]
Comment (sorahn) [reply api-proposal.md]: It should probably be its own entry.

File: api-proposal.md (hunk 3 above)
Comment (sorahn) [reply api-proposal.md]: Yeah I'm not sure what needs to go in here yet, But I will ping it, and if it 400's or something, they're not logged in, and if it returns a username, they are.

Probably some other github related things. avatar? userid? I don't actually know what github oauth will give you.

Comment (klion26) [reply api-proposal.md]: Yes, if we need the other things GitHub returns, we can add it later.

**Comment by rmetzger** [comment]
I would like to propose some minor changes to this. How about we merge this as is, and I open another PR?

File: api-proposal.md (hunk 3 above)
Comment (klion26) [reply api-proposal.md]: After another think, do we need to check the authentication in the backend? or we just need to it in the frontend?

When we hit the "login with github" on the [page](https://flink-ecosystem-demo.flink-resources.org/#/package/incredible-fresh-hat), we will not go to the backend.

Comment (sorahn) [reply api-proposal.md]: Github's oAuth setup does not allow for a front-end only solution.

**Comment by rmetzger** [comment]
Thanks for merging.

**Pull request merged**